### DISCUSSION TITLE
Dealing with hyphens in StandardAnalyzer?

### DISCUSSION THREAD

**Discussion by kamranayub** [ravendb/ravendb#18214] [Q&A]
For the Standard Analyzer in the docs it says it will treat hyphens with numbers as a "product code".

> [Docs:](https://ravendb.net/docs/article-page/6.0/csharp/indexes/using-analyzers#ravendb) Splits words at hyphens, unless there's a number in the token, in which case the whole token is interpreted as a product number and is not split.

When indexing "slugified" terms this messes with tokenization. For example, in my DB I've got a slugified title like this:

`half-life-2-lost-coast`

And I would expect it to be tokenized like this:

`half, life, 2, lost, coast`

But using the RavenStandardAnalyzer, I see this:

`coast, half, life-2-lost`

So if I do a query like `search(SlugifiedName, 'half-life-2')` it doesn't return any results which is obviously not what I want. With games and movies especially, sequels are common and numbers matter, so I don't want to use the other analyzers that strip numbers. 

Since this is customized using the Lucene.Net packaged with RavenDB, I'm wondering what the easiest way is to restore the "original" functionality (or otherwise disable this one extra nuance for product codes).

Do I have to provide the original/latest Lucene.Net assembly somehow (so it doesn't overwrite the shipped version)?

**Edit:** I was confused initially. Hyphens + alpha numeric handling is part of the original StandardAnalyzer. See the `NUM` definition here:

https://github.com/apache/lucenenet/blob/Lucene.Net_3_0_3_RC2_final/src/core/Analysis/Standard/StandardTokenizerImpl.jflex#L120-L125

ChatGPT is actually helpful here, and this is what it explained:

> Here's a breakdown of the NUM pattern as defined:
>
> ```
> NUM = ({ALPHANUM} {P} {HAS_DIGIT}
>      | {HAS_DIGIT} {P} {ALPHANUM}
>      | {ALPHANUM} ({P} {HAS_DIGIT} {P} {ALPHANUM})+
>      | {HAS_DIGIT} ({P} {ALPHANUM} {P} {HAS_DIGIT})+
>      | {ALPHANUM} {P} {HAS_DIGIT} ({P} {ALPHANUM} {P} {HAS_DIGIT})+
>      | {HAS_DIGIT} {P} {ALPHANUM} ({P} {HAS_DIGIT} {P} {ALPHANUM})+)
> ```
>
> In this pattern:
> 
> `{ALPHANUM}` stands for a sequence of alphabetic (letter) or numeric (digit) characters.
> `{P}` represents punctuation characters defined as `("_"|"-"|"/"|"."|",")`.
> `{HAS_DIGIT}` is a sequence that must contain at least one digit.
> According to this definition:
> 
> The NUM rule accommodates sequences where alphabetic and numeric elements are interspersed with punctuation, which includes hyphens (-). This rule is specifically designed to capture complex alphanumeric patterns like serial numbers, model numbers, or IP addresses, where numbers and letters can be mixed and separated by punctuation such as hyphens.
> Applying this to `half-life-2-lost-coast`:
> 
> The segment `life-2-lost` matches one of these patterns because it is an alphanumeric segment (`{ALPHANUM}`), followed by punctuation ({P}, the hyphen in this case), followed by another alphanumeric segment that contains a digit (`{HAS_DIGIT}`), which in your example is "2". Therefore, according to the NUM pattern,` life-2-lost` stays as a single token instead of being split at the hyphens.
> The presence of numbers mixed with hyphens and alphabetic characters has caused the tokenizer to treat "life-2-lost" as a single entity, likely because it fits the logic designed to capture more complex alphanumeric codes or identifiers. This is a more specialized handling than a simple split on all punctuation, reflecting the nuanced approach required for certain types of content like technical documents, serial numbers, or codes where such patterns are meaningful.
> 
> Understanding this helps explain why the tokenizer does not split the string into separate tokens at every hyphen. This specific behavior is intentional, designed to preserve certain types of information that could be lost with a more naive tokenization approach.

It actually handles other cases like serial numbers and such. So in my case, I'm indexing a slug with hyphens, and it gets treated as such.

**Comment by kamranayub** [comment]
So I think one approach I can do is to index slugified titles as terms without hyphens. That will allow the Standard Analyzer to split the terms normally. When I do that, I get what I expect and I can still index the exact slug to boost exact matches above searches.

**Comment by ayende** [comment]
As you noted, this is the standard behavior of the `StandardAnalyzer`. There are two ways to handle this:

1. Easiest, do the splitting directly, in your index: `Slug = item.Slug.Split('-')`. This gives you the most control and matches exactly what you are trying to do.
2. Use a custom analyzer, see: https://ayende.com/blog/200833-B/using-analyzers-in-ravendb-indexing
